chain-signatures/node/src/protocol/posit.rs

```diff
@@ -0,0 +1,270 @@
+                // Checks:
+                // 1. We are not the proposer.
+                // 2. Somebody else hasn't also proposed the protocol.
+
+                if let Some(positor) = self.posits.get(&id) {
+                    let proposer = positor.id();
+                    if positor.is_proposer() {
+                        tracing::warn!(?id, ?from, "received INIT on protocol we already proposed");
+                        return PositInternalAction::Reply(PositAction::Reject);
+                    } else if proposer != from {
+                        tracing::warn!(
+                            ?id,
+                            ?from,
+                            ?proposer,
+                            "received INIT on conflicting proposer"
+                        );
+                        return PositInternalAction::Reply(PositAction::Reject);
+                    }
+                } else {
+                    self.posits.put(id, Positor::Deliberator(from));
+                }
+
+                // No further checks necessary, we can just accept the posit.
+                PositInternalAction::Reply(PositAction::Accept)
+            }
+            PositAction::Start(participants) => {
+                // Checks:
+                // 1. We are a participant in the protocol.
+                // 2. We are not the proposer.
+                // 3. The proposer is the one that started the protocol.
+
+                if !participants.contains(&self.me) {
+                    tracing::warn!(
+                        ?id,
+                        ?from,
+                        "received START on protocol we are not a part of"
+                    );
+                    return PositInternalAction::Reply(PositAction::Reject);
+                }
+
+                if let Some(positor) = self.posits.pop(&id) {
+                    let proposer = positor.id();
+                    if positor.is_proposer() {
+                        tracing::warn!(
+                            ?id,
+                            ?from,
+                            "received START on protocol we already proposed"
+                        );
+                        self.posits.put(id, positor);
+                        return PositInternalAction::Reply(PositAction::Reject);
+                    } else if proposer != from {
+                        tracing::warn!(
+                            ?id,
+                            ?from,
+                            ?proposer,
+                            "received START on conflicting proposer"
+                        );
+                        self.posits.put(id, positor);
+                        return PositInternalAction::Reply(PositAction::Reject);
+                    }
+                } else {
+                    tracing::warn!(?id, ?from, "received START on protocol we have no info for");
+                    return PositInternalAction::Reply(PositAction::Reject);
+                }
```

Comment:
Just checking: Is this handling the case where we get a `Start` message before the `Propose` message?

Reply:
yep, the node does not have any info about this posit, so that would mean no proposer has sent out a propose yet